packages/forms/src/patterns/social-security-number/social-security-number.ts

```diff
@@ -0,0 +1,102 @@
+import * as z from 'zod';
+import { type SocialSecurityNumberProps } from '../../components.js';
+import { type Pattern, type PatternConfig } from '../../pattern.js';
+import { getFormSessionValue } from '../../session.js';
+import {
+  safeZodParseFormErrors,
+  safeZodParseToFormError,
+} from '../../util/zod.js';
+
+const configSchema = z.object({
+  label: z.string().min(1),
+  required: z.boolean(),
+  hint: z.string().optional(),
+});
+
+export type SocialSecurityNumberPattern = Pattern<z.infer<typeof configSchema>>;
+
+export type SocialSecurityNumberPatternOutput = z.infer<
+  ReturnType<typeof createSSNSchema>
+>;
+
+export const createSSNSchema = (data: SocialSecurityNumberPattern['data']) => {
+  const ssnSchema = z.string().refine(value => {
+    const stripped = value.replace(/[^0-9]/g, '');
+
+    const isValidFormat = /^\d{3}-\d{2}-\d{4}$|^\d{9}$/.test(value);
+
+    if (!isValidFormat) {
+      return false;
+    }
+
+    if (stripped.length !== 9) {
+      return false;
+    }
+
+    const hasValidPrefix = !(
+      stripped.startsWith('9') ||
+      stripped.startsWith('666') ||
+      stripped.startsWith('000')
+    );
+
+    const hasValidMiddle = stripped.slice(3, 5) !== '00';
+    const hasValidSuffix = stripped.slice(5) !== '0000';
+
+    return hasValidPrefix && hasValidMiddle && hasValidSuffix;
+  }, 'Social Security Number must contain exactly 9 digits, be formatted as XXX-XX-XXXX or XXXXXXXXX, and meet SSA issuance criteria');
```

Comment:
We might be able to validate this in user testing, but it may be better to break up the checks here to provide a more targeted error message. 

Zod allows you to chain validation rules with its `refine` function, and it'll return the error message of the first rule that returns `false`. You could do something like:

```
z.string()
  .transform(value => // return the cleaned string)
  .refine(value => !(data.required && value.length === 0), 'This field is required')
  .refine(value => // check for length, 'Error message for the correct amount of characters')
  .refine(value => // check for valid pattern, 'Error message for invalid number')
```

I'd encourage the use of the `transform` function especially because that will help normalize how the data is stored.

You would also be able to shed the `.or` operator in the line below if you did it this way.

Reply:
Also when I have multiple .refines() I am getting type errors for nesting:

```
Type 'ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>, string, string>, string, string>, string, string>' is not assignable to type 'ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>, string, string>, string, string>'.
  Type 'ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>, string, string>, string, string>' is not assignable to type 'ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>, string, string>'.
    Type 'ZodEffects<ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>, string, string>' is not assignable to type 'ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>'.
      Type 'ZodEffects<ZodEffects<ZodEffects<ZodString, string, string>, string, string>, string, string>' is not assignable to type 'ZodEffects<ZodEffects<ZodString, string, string>, string, string>'.
        Type 'ZodEffects<ZodEffects<ZodString, string, string>, string, string>' is not assignable to type 'ZodEffects<ZodString, string, string>'.
          Type 'ZodEffects<ZodString, string, string>' is missing the following properties from type 'ZodString': _regex, _addCheck, email, url, and 39 more.ts(2322)
let schema: z.ZodEffects<z.ZodEffects<z.ZodEffects<z.ZodEffects<z.ZodEffects<z.ZodString, string, string>, string, string>, string, string>, string, string>, string, string>
```

Reply:
However, I can refactor code to this code block with better error messages:

```
export const createSSNSchema = (data: SocialSecurityNumberPattern['data']) => {
  const baseSchema = z.string()
    .transform(value => value.replace(/[^0-9]/g, ''))
    .superRefine((value, ctx) => {
      let issues = [];

      if (value.length !== 9) {
        issues.push('exactly 9 digits');
      } else {
        if (value.startsWith('9') || value.startsWith('666') || value.startsWith('000')) {
          issues.push('an invalid prefix');
        }

        if (value.slice(3, 5) === '00') {
          issues.push('an invalid middle segment');
        }

        if (value.slice(5) === '0000') {
          issues.push('an invalid suffix');
        }
      }

      if (issues.length > 0) {
        let enhancedMessage = `Social Security Number must have `;
        if (issues.length === 1) {
          enhancedMessage += issues[0];
        } else if (issues.length === 2) {
          enhancedMessage += `${issues[0]} and ${issues[1]}`;
        } else {
          enhancedMessage += `${issues.slice(0, -1).join(', ')}, and ${issues[issues.length - 1]}`;
        }
        
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: enhancedMessage,
        });
      }
    });

  if (data.required) {
    return z.string()
      .refine(value => value.trim().length > 0, {
        message: 'This field is required',
      })
      .superRefine((value, ctx) => {
        const result = baseSchema.safeParse(value);
        if (!result.success) {
          result.error.issues.forEach(issue =>
            ctx.addIssue(issue)
          );
        }
      });
  } else {
    return baseSchema.optional();
  }
};
```

<img width="824" alt="Screenshot 2024-11-21 at 2 25 27 PM" src="https://github.com/user-attachments/assets/b9844ade-c0f4-410e-b0b9-db5b17d5fdb7">

<img width="815" alt="Screenshot 2024-11-21 at 2 25 14 PM" src="https://github.com/user-attachments/assets/84e88b9f-41a3-4ed9-a68c-64a4abd43a00">

<img width="832" alt="Screenshot 2024-11-21 at 2 25 02 PM" src="https://github.com/user-attachments/assets/8d429907-87dd-4680-8050-942b4c561d3f">

Reply:
I can refactor the code block to:

```
export const createSSNSchema = (data: SocialSecurityNumberPattern['data']) => {
  const baseSchema = z.string()
    .transform(value => value.replace(/[^0-9]/g, '')) // Normalize the SSN by removing non-numeric characters
    .superRefine((value, ctx) => {
      let issues = [];

      if (value.length !== 9) {
        issues.push('have exactly 9 digits');
      } else {
        if (value.startsWith('9') || value.startsWith('666') || value.startsWith('000')) {
          issues.push('start with a valid prefix (not 9, 666, or 000)');
        }

        if (value.slice(3, 5) === '00') {
          issues.push('have a valid middle segment (not 00)');
        }

        if (value.slice(5) === '0000') {
          issues.push('have a valid suffix (not 0000)');
        }
      }

      if (issues.length > 0) {
        let enhancedMessage = 'Social Security Number must ';
        if (issues.length === 1) {
          enhancedMessage += issues[0];
        } else if (issues.length === 2) {
          enhancedMessage += `${issues[0]} and ${issues[1]}`;
        } else {
          enhancedMessage += `${issues.slice(0, -1).join(', ')}, and ${issues[issues.length - 1]}`;
        }
        
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: enhancedMessage,
        });
      }
    });

  if (data.required) {
    return z.string()
      .refine(value => value.trim().length > 0, {
        message: 'This field is required',
      })
      .superRefine((value, ctx) => {
        const result = baseSchema.safeParse(value.trim());
        if (!result.success) {
          result.error.issues.forEach(issue =>
            ctx.addIssue(issue)
          );
        }
      });
  } else {
    return baseSchema.optional();
  }
};
```

<img width="753" alt="Screenshot 2024-11-21 at 2 36 10 PM" src="https://github.com/user-attachments/assets/d690b604-046a-48f6-817b-44f6a9c43f9f">

<img width="748" alt="Screenshot 2024-11-21 at 2 36 00 PM" src="https://github.com/user-attachments/assets/25340a3d-89f0-4b31-899e-8bcbe9a71ca9">

<img width="758" alt="Screenshot 2024-11-21 at 2 35 46 PM" src="https://github.com/user-attachments/assets/6a0defa7-a858-4044-8f32-1d05ccb1005f">

<img width="748" alt="Screenshot 2024-11-21 at 2 35 27 PM" src="https://github.com/user-attachments/assets/b6e330cb-5d0a-4bae-93bb-9e4d780be2c4">

<img width="770" alt="Screenshot 2024-11-21 at 2 35 06 PM" src="https://github.com/user-attachments/assets/849730f1-b748-4eef-a726-4e743ef2ed1a">

Reply:
Thanks for taking a look at this. My concern before was that the error was complex with all the criteria that a valid SSN had to meet. I think for now, we can call this good enough for our current epic, but we might want to get some validation on what's the right balance of helping the user figure out why their input is invalid vs. cognitive load.

I think the version you have with `.superRefine` is good for now. If you wouldn't mind, can you create a design issue for the next epic so we can get some testing on the language?

Reply:
Created the ticket: https://github.com/GSA-TTS/forms/issues/389